lib/c-api/src/wasm_c_api/unstable/target_lexicon.rs

```diff
@@ -0,0 +1,318 @@
+///     assert(triple);
+///
+///     wasm_triple_delete(triple);
+///
+///     return 0;
+/// }
+/// #    })
+/// #    .success();
+/// # }
+/// ```
+///
+/// See also [`wasm_triple_new`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_triple_new_from_host() -> Box<wasm_triple_t> {
+    Box::new(wasm_triple_t {
+        inner: Triple::host(),
+    })
+}
+
+/// Delete a [`wasm_triple_t`].
+///
+/// # Example
+///
+/// See [`wasm_triple_t`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_triple_delete(_triple: Option<Box<wasm_triple_t>>) {}
+
+/// Unstable non-standard Wasmer-specific API to represent a set of
+/// CPU features.
+///
+/// CPU features are identified by their stringified names. The
+/// reference is the GCC options:
+///
+/// * <https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html>,
+/// * <https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html>,
+/// * <https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html>.
+///
+/// At the time of writing this documentation (it might be outdated in
+/// the future), the supported features are the following:
+///
+/// * `sse2`,
+/// * `sse3`,
+/// * `ssse3`,
+/// * `sse4.1`,
+/// * `sse4.2`,
+/// * `popcnt`,
+/// * `avx`,
+/// * `bmi`,
+/// * `bmi2`,
+/// * `avx2`,
+/// * `avx512dq`,
+/// * `avx512vl`,
+/// * `lzcnt`.
+///
+/// # Example
+///
+/// ```rust
+/// # use inline_c::assert_c;
+/// # fn main() {
+/// #    (assert_c! {
+/// # #include "tests/wasmer_wasm.h"
+/// #
+/// int main() {
+///     // Create a new CPU feature set.
+///     wasm_cpu_features_t* cpu_features = wasm_cpu_features_new();
+///
+///     // Create a new feature name, here `sse2`, and add it to the set.
+///     {
+///         wasm_name_t cpu_feature_name;
+///         wasm_name_new_from_string(&cpu_feature_name, "sse2");
+///
+///         wasm_cpu_features_add(cpu_features, &cpu_feature_name);
+///
+///         wasm_name_delete(&cpu_feature_name);
+///     }
+///
+///     wasm_cpu_features_delete(cpu_features);
+///
+///     return 0;
+/// }
+/// #    })
+/// #    .success();
+/// # }
+/// ```
+#[allow(non_camel_case_types)]
+pub struct wasm_cpu_features_t {
+    inner: EnumSet<CpuFeature>,
+}
+
+/// Create a new [`wasm_cpu_features_t`].
+///
+/// # Example
+///
+/// See [`wasm_cpu_features_t`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_cpu_features_new() -> Box<wasm_cpu_features_t> {
+    Box::new(wasm_cpu_features_t {
+        inner: CpuFeature::set(),
+    })
+}
+
+/// Delete a [`wasm_cpu_features_t`].
+///
+/// # Example
+///
+/// See [`wasm_cpu_features_t`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_cpu_features_delete(_cpu_features: Option<Box<wasm_cpu_features_t>>) {
+}
+
+/// Add a new CPU feature into the set represented by
+/// [`wasm_cpu_features_t`].
+///
+/// # Example
+///
+/// See [`wasm_cpu_features_t`].
+#[no_mangle]
+pub unsafe extern "C" fn wasm_cpu_features_add(
+    cpu_features: Option<&mut wasm_cpu_features_t>,
+    feature: Option<&wasm_name_t>,
```

Comment:
I don't think we need `Option` around all user-supplied pointers, the standard wasm-c-api doesn't bother with that in a lot of cases, if the pointer is invalid it'll crash or do what C does

Reply:
Yeah but I prefer to prevent bad usage of our API. That's what we already do everywhere almost.